pkg/deploy/aga/listener_synthesizer.go

```diff
@@ -0,0 +1,532 @@
+			"portRanges", s.portRangesToString(resListener.Spec.PortRanges))
+
+		listenerStatus, err := s.listenerManager.Create(ctx, resListener)
+		if err != nil {
+			// If we hit a listener limit error, log it clearly
+			var apiErr *agatypes.LimitExceededException
+			if errors.As(err, &apiErr) {
+				s.logger.Error(err,
+					"Reached listener limit on accelerator. Tried to create a listener after deleting unmatched ones.")
+			}
+			return err
+		}
+		resListener.SetStatus(listenerStatus)
+	}
+
+	return nil
+}
+
+// findSDKListenersOnAccelerator returns all listeners for the given accelerator
+func (s *listenerSynthesizer) findSDKListenersOnAccelerator(ctx context.Context, accARN string) ([]*ListenerResource, error) {
+	// List listeners for the accelerator
+	listInput := &globalaccelerator.ListListenersInput{
+		AcceleratorArn: awssdk.String(accARN),
+	}
+	sdkListeners, err := s.gaClient.ListListenersAsList(ctx, listInput)
+	if err != nil {
+		var apiErr *agatypes.AcceleratorNotFoundException
+		if errors.As(err, &apiErr) {
+			s.logger.Info("Accelerator not found in AWS, skipping listener listing",
+				"acceleratorARN", accARN)
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to list listeners for accelerator %s", accARN)
+	}
+
+	var listeners []*ListenerResource
+	for _, listener := range sdkListeners {
+		// Clone the listener as the range variable address is reused
+		listenerCopy := listener
+		listeners = append(listeners, &ListenerResource{
+			Listener: &listenerCopy,
+		})
+	}
+	return listeners, nil
+}
+
+// resAndSDKListenerPair holds a matched pair of resource and SDK listener
+type resAndSDKListenerPair struct {
+	resListener *agamodel.Listener
+	sdkListener *ListenerResource
+}
+
+// matchResAndSDKListeners matches resource listeners with SDK listeners using a similarity-based algorithm
+// that finds the best match based on protocol and port range overlap.
+// Returns three groups:
+// - matchedResAndSDKListeners: pairs of resource and SDK listeners that will be updated
+// - unmatchedResListeners: resource listeners that don't match any SDK listeners and will be created
+// - unmatchedSDKListeners: SDK listeners that don't match any resource listeners and will be deleted
+func (s *listenerSynthesizer) matchResAndSDKListeners(resListeners []*agamodel.Listener, sdkListeners []*ListenerResource) (
+	[]resAndSDKListenerPair, []*agamodel.Listener, []*ListenerResource) {
+
+	// First, try to match by exact protocol and port ranges
+	exactMatches, remainingResListeners, remainingSDKListeners := s.findExactMatches(resListeners, sdkListeners)
+
+	// For remaining listeners, use similarity-based matching
+	similarityMatches, unmatchedResListeners, unmatchedSDKListeners := s.findSimilarityMatches(
+		remainingResListeners, remainingSDKListeners)
+
+	// Combine exact and similarity matches
+	matchedPairs := append(exactMatches, similarityMatches...)
+
+	s.logger.V(1).Info("Matched listeners",
```

Comment:
It might be useful to log the key+arn for the exact and similarity matches here? That way, a user could understand why a given listener was updated vs. deleted.